### PR TITLE
ledger: handle machine exceptions

### PR DESCRIPTION
**Pull request by michaelpj**
These should certainly fail validation.

Arguably, the validating node should just be catching all kinds of
exceptions and treating them as failures. I don't think this is right:
the exceptions we get here are *non-transient* errors determined by the
program we've been given, rather than e.g. ambient machine conditions.
So these really do indicate that the script is bad, whereas other
exceptions may not.

As mentioned in https://github.com/input-output-hk/plutus/issues/1524, this is more relevant if we start erasing things.

### CODE REVIEW

File: plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs [hunk 1]
```diff
@@ -224,6 +225,11 @@ readKnownCek means term = do
             appErr = ConstAppMachineError $ UnreadableBuiltinConstAppError term err
         EvaluationSuccess (Right x)  -> Right $ EvaluationSuccess x
 
+-- | Run a program using the CEK machine.
+-- Calls 'evaluateCekCatch' under the hood.
+runCekCatch :: DynamicBuiltinNameMeanings -> Program TyName Name () -> Either CekMachineException EvaluationResultDef
+runCekCatch means (Program _ _ term) = evaluateCekCatch means term
```
Comment (effectfully) [inline plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs]: If this is now a default, then I'd rename `runCek` to `unsafeRunCek` and `runCekCatch` to `runCek`.

File: plutus-tx/test/TH/Spec.hs [hunk 2]
```diff
@@ -44,13 +44,16 @@ runPlcCek :: GetProgram a => [a] -> ExceptT SomeException IO EvaluationResultDef
 runPlcCek values = do
      ps <- Haskell.traverse getProgram values
      let p = foldl1 applyProgram ps
-     ExceptT $ try @SomeException $ evaluate $ evaluateCek p
+     let result = evaluateCek p
```
Comment (effectfully) [inline plutus-tx/test/TH/Spec.hs]: `-Wall` doesn't say this is a redundant line?

File: plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs]: Hm yes.

File: plutus-tx/test/TH/Spec.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-tx/test/TH/Spec.hs]: :man_facepalming:

**Pull request merged**